Fix #3134: Admin Settings shows error message for null pin entries.

Arnold2381 opened the pull request:
## Explanation
Fixes #3134 

https://user-images.githubusercontent.com/46752548/117769644-96fa1900-b251-11eb-86f0-9363c7067383.mp4

Screenshot of all the tests passing on espresso
![image](https://user-images.githubusercontent.com/46752548/126890560-a7e214ca-c436-4b8c-b18c-6d6f3f212a10.png)

![image](https://user-images.githubusercontent.com/46752548/119657086-19d1c500-be49-11eb-81ad-f266de073055.png)



## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Arnold2381 committed:
Added an error message if the admin Pin field is submitted with Null string

anandwana001 commented:
Write test case too. 

https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing

Arnold2381 commented:
@anandwana001 I have a question, in the Administrator Controls, the submit button only gets activated when the user enters a pin of length 5 but in this PR in the video if you see when a user is going to Administrator controls even if the submit button isn't activated, he can use his keyboard submit button and if he does that with an incorrect pin of whatever length less than 5 an error is shown, so I did the same for null string but this cant be tested. So is it okay if I don't include tests for this?

Arnold2381 committed:
Tests added for null pin entries in Access to Adminstrator Settings.

anandwana001 commented:
@Arnold2381 Can we disable the keyboard from submitting?  this makes out the keyboard in sync with the submit button state.

Arnold2381 commented:
> @Arnold2381 Can we disable the keyboard from submitting? this makes out the keyboard in sync with the submit button state.

I did figure out a way to test it. But yes we can remove it, and if I do that, I will be removing the tests for that.

Arnold2381 committed:
Disabled numpad submit buttton.

Arnold2381 committed:
Merge branch 'develop' into my-admin-branch

Arnold2381 committed:
Removed an unused import.

Arnold2381 commented:
@anandwana001 Also PTAL in this one.

anandwana001 commented on app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt:
```diff
@@ -58,8 +57,7 @@ class AdminAuthActivityPresenter @Inject constructor(
     binding.adminAuthInputPinEditText.setOnEditorActionListener { _, actionId, event ->
-      if (actionId == EditorInfo.IME_ACTION_DONE ||
-        (event != null && (event.keyCode == KeyEvent.KEYCODE_ENTER))
+      if (event != null && (event.keyCode == KeyEvent.KEYCODE_ENTER)
```
Are we able to use the keyboard action done,  if we input the correct pin in the input field? or Do we have to only use the submit button?

Arnold2381 replied:
Yes, we can use the keyboard button. It same as the submit button now.
Until and unless the length of the pin entered is less than the adminPin's length the keyboard submit can be clicked but would have no action.

anandwana001 replied:
> but would have no action.

Will this show any error message or nothing will happen?

Arnold2381 replied:
No, it won't show. It will act as ditto as the submit button, so nothing will happen.

anandwana001 commented on app/src/main/res/values/strings.xml:
```diff
@@ -201,6 +201,7 @@
   <string name="admin_auth_admin_controls_sub">Enter the Administrator PIN in order to access Administrator Controls.</string>
   <string name="admin_auth_pin">Administrator\'s PIN</string>
   <string name="admin_auth_incorrect">Incorrect Administrator PIN. Please try again.</string>
+  <string name="admin_auth_null">This field can\'t be empty.</string>
```
@rt4914 Do we need to confirm this message with Chantel or does it sounds good?

rt4914 replied:
@mschanteltc If you check the video in the issue: https://github.com/oppia/oppia-android/issues/3134 you will see that if the learner clicks on `Submit` button if there is no text then there is no error message. Unfortunately we cannot `disable` the `Submit` button here so we will have to show error message only. 
So does `This field can't be empty` makes sense to me?
To me `Please enter Administrator Pin` makes more sense meaning something precise as an error.
WDYT?

mschanteltc replied:
I agree that the latter choice `Please enter Administrator PIN` works better.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt:
```diff
@@ -216,30 +191,6 @@ class AdminAuthActivityTest {
   }
 
-  @Test
-  fun testAdminAuthActivity_inputCorrectPassword_imeAction_opensAdminControlsActivity() {
```
I don't think it's an appropriate solution to solve this issue, we do need the keyboard action done feature so that users can submit input using the keyboard also, not dependent on the submit button.

Arnold2381 replied:
Yes, I have changed it now. 
The keyboard button can be used to submit but only when the length of the pin entered is the same as the adminPin's length.

anandwana001 replied:
Do we need to test in `app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt` as well? 
As we are removing the test from here, I think we should update the test as per the button/keyboard thing rather than removing the test.

Arnold2381 replied:
I made some change now, and I don't think we need those tests, the two tests I removed are testing cases which won't occur now. A pin less than the length of the admin pin will have no action when the keyboard submit button is pressed. So I don't think they are needed.
Your views?

anandwana001 reviewed (changes requested):
Thanks @Arnold2381 

left a few thoughts, will check test file in next iteration.

Arnold2381 committed:
Keyboard submit button insync with the submit button.

Arnold2381 committed:
Keyboard submit button insync with the submit button.

Arnold2381 commented:
@anandwana001 Sorry for the late reply, I was busy with my college exams and everything.
I have made the keyboard button in sync with the submit button. And also user can use the keyboard submit button only when he enters a pin of length equal to the admin pin. And also because we are removing the functionality of not allowing the user to press the submit button before the admin pin's length, I don't think we need the tests for that so I removed the two tests for the same.
PTAL

Arnold2381 committed:
Merge branch 'oppia:develop' into my-admin-branch

anandwana001 commented:
@Arnold2381 Please follow up on all my earlier comments.
Also, Chantel has confirm the string here - https://github.com/oppia/oppia-android/pull/3181#discussion_r640158432

Arnold2381 committed:
Keyboard submit button insync with the submit button.

anandwana001 commented:
Hi @Arnold2381 

Looks like we lost track of this PR, are you still working on this, and could you update on where are we on this PR. Also, whenever you need a review of any member, please comment the persons' GitHub handle and add `PTAL`, our Oppiabot will assign that person to the assignee section so that the PR doesn't go off the track.

Arnold2381 commented:
Hello @anandwana001 
Sorry! I thought replying to your reviews would notify you, next time I would do the same. .

Yea so our discussion was on the tests I removed. I suggest there is no need of the tests and I removed it because the cases its testing for will never arise.  As you will see when will enter a pin less then the length of the adminPin the button doesn't gets activated so I don't think it will ever show an error. The tests were required before because the keyboard submit button was allowing the user to press the submit button before the length of the admin pin entered but now, the keyboard submit button will have no action until and unless the pin length is same as admin and once same it acts same as the normal submit button.
So i think whatever is done till now is working.

Arnold2381 commented:
@anandwana001 PTAL

rt4914 commented:
@anandwana001 PTAL at this PR.

anandwana001 commented:
I will take a look at this tonight.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt:
```diff
@@ -291,75 +291,6 @@ class AdminAuthActivityTest {
   }
 
-  @Test
-  fun testAdminAuthActivity_inputIncorrectPassword_inputAgain_errorIsGone() {
```
why are we removing this test?

Arnold2381 replied:
When soft keyboard didn't function as same as the submit button we needed this test. because that time even if the length of the pin was less then the length of the admin pin the soft keyboard submit was working but now as I have already added the functionality that soft keyboard submit button will only work when the length is same for both entered pin and admin pin there is no use of this.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt:
```diff
@@ -291,75 +291,6 @@ class AdminAuthActivityTest {
-  }
-
-  @Test
-  fun testAdminAuthActivity_inputIncorrectPassword_correct_imeAction_errorIsGone() {
```
why are we removing this test?

Arnold2381 replied:
same.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -673,6 +673,73 @@ class PinPasswordActivityTest {
   }
 
+  @Test
+  fun testPinPassword_withUser_forgot_inputAdminPinAndNullPin_errorIsDisplayed() {
```
Also, add a configChange version on this test and separately test it.

Arnold2381 replied:
Done and updated the sc of passed tests.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -673,6 +673,73 @@ class PinPasswordActivityTest {
+        )
+      ).inRoot(isDialog())
+        .perform(editTextInputAction.appendText(""), closeSoftKeyboard())
+      onView(withText(context.getString(R.string.admin_settings_submit)))
```
Add a test case where click on the action done in softKeyboard, rather than `admin_settings_submit`.

Arnold2381 replied:
ditto.

anandwana001 reviewed (changes requested):
Tested it, seems working for me. 

Nit queries regarding test cases, else LGTM.

Arnold2381 committed:
Added tests for input admin nullPin.

Arnold2381 commented:
@anandwana001 PTAL.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -673,6 +674,147 @@ class PinPasswordActivityTest {
+  }
+
+  @Test
+  fun testPinPassword_User_forgot_adminPinAndNullPin_configChange_imeAction_errorIsDisplayed() {
```
Rename to `testPinPassword_user_forgot_adminPinAndNullPin_configChange_imeAction_errorIsDisplayed`

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -673,6 +674,147 @@ class PinPasswordActivityTest {
+          isDescendantOfA(withId(R.id.admin_settings_input_pin))
+        )
+      ).inRoot(isDialog())
+        .perform(editTextInputAction.appendText(""), ViewActions.pressImeActionButton())
```
Use direct imports
`.perform(editTextInputAction.appendText(""), pressImeActionButton())`

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -673,6 +674,147 @@ class PinPasswordActivityTest {
+          isDescendantOfA(withId(R.id.admin_settings_input_pin))
+        )
+      ).inRoot(isDialog())
+        .perform(editTextInputAction.appendText(""), ViewActions.pressImeActionButton())
```
Ditto

rt4914 reviewed (approved):
Nit changes suggested. Else looks good. 

Merge with latest develop and wait for @anandwana001 approval.

Arnold2381 committed:
Few nits and imports fixed.

Arnold2381 committed:
Merge branch 'oppia:develop' into my-admin-branch

Arnold2381 commented:
@rt4914 done!

rt4914 reviewed (approved):
LGTM, thanks.

anandwana001 reviewed (approved):
LGTM
